x/ccv/provider/keeper/misbehaviour.go

```diff
@@ -0,0 +1,166 @@
+package keeper
+
+import (
+	"fmt"
+
+	ibcclienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	ibctmtypes "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint"
+
+	errorsmod "cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	tmtypes "github.com/cometbft/cometbft/types"
+
+	"github.com/cosmos/interchain-security/v3/x/ccv/provider/types"
+)
+
+// HandleConsumerMisbehaviour checks if the given IBC misbehaviour corresponds to an equivocation light client attack,
+// and in this case, slashes, jails, and tombstones the Byzantine validators
+func (k Keeper) HandleConsumerMisbehaviour(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) error {
+	logger := k.Logger(ctx)
+
+	// Check that the misbehaviour is valid and that the client consensus states at trusted heights are within trusting period
+	if err := k.CheckMisbehaviour(ctx, misbehaviour); err != nil {
+		logger.Info("Misbehaviour rejected", err.Error())
+
+		return err
+	}
+
+	// Since the misbehaviour packet was received within the trusting period
+	// w.r.t to the trusted consensus states the infraction age
+	// isn't too old. see ibc-go/modules/light-clients/07-tendermint/types/misbehaviour_handle.go
+
+	// Get Byzantine validators from the conflicting headers
+	byzantineValidators, err := k.GetByzantineValidators(ctx, misbehaviour)
+	if err != nil {
+		return err
+	}
+
+	provAddrs := make([]types.ProviderConsAddress, len(byzantineValidators))
+
+	var errors []error
+	// slash, jail, and tombstone the Byzantine validators
+	for _, v := range byzantineValidators {
+		providerAddr := k.GetProviderAddrFromConsumerAddr(
+			ctx,
+			misbehaviour.Header1.Header.ChainID,
+			types.NewConsumerConsAddress(sdk.ConsAddress(v.Address.Bytes())),
+		)
+		err := k.SlashValidator(ctx, providerAddr)
+		if err != nil {
+			errors = append(errors, err)
+		}
+		err = k.JailAndTombstoneValidator(ctx, providerAddr)
+		if err != nil {
+			errors = append(errors, err)
+		}
+		provAddrs = append(provAddrs, providerAddr)
+	}
+
+	logger.Info(
+		"confirmed equivocation light client attack",
+		"byzantine validators", provAddrs,
+	)
+
+	// If we fail to slash all validators we return an error. However, if we only fail to slash some validators
+	// we just log an error to avoid having the whole `MsgSubmitMisbehaviour` failing and reverting the partial slashing.
+	if len(errors) == len(byzantineValidators) {
```

Comment:
Here, it should've been:
```suggestion
	if len(errors) == 2 * len(byzantineValidators) {
```
because an error is appended once when `SlashValidator` fails and a second time when `JailAndTombstoneValidator` fails. My mistake.

Technically, `len(errors)` should always be an even number, otherwise something went wrong (according to my previous [comment](https://github.com/cosmos/interchain-security/pull/1340/files#r1356584667)). Maybe we can log something if that's not the case -- it's not clear to me who's even looking at the logs though? Is the assumption here that when we log something, eventually a validator would ping us about them?